### PR TITLE
Custom resource actions should be available by default

### PR DESCRIPTION
**Pull request by alexmt**

### CODE REVIEW

**Comment by simster7** [comment]
@alexmt Curious – why is this change being made?

**Comment by alexmt** [comment]
I should've asked for your review @simster7 . I've noticed that I cannot use restart action on deployment because it is not available.

It appears that after merging https://github.com/argoproj/argo-cd/pull/2110 all actions are not available unless the available field is set to true. One way to fix it is to update all discovery.lua . In most cases, actions are always available so instead, I've decided to make action available by default and it can be made unavailable by adding field `availble = false`.

Does it makes sense?

**Comment by simster7** [comment]
@alexmt Hmmm, we don't check if an action is available before running it. "Available" is purely an informative label when running `argocd app actions APPNAME list`. So even if the "Available" label returns `false`, the action _should_ still run correctly if it is indeed available.

**Comment by simster7** [comment]
> One way to fix it is to update all discovery.lua

Yes, @jessesuen and I knew this would have to be done when merging.

> In most cases, actions are always available so instead, I've decided to make action available by default and it can be made unavailable by adding field `availble = false`

In this case, then I think the correct way to go is to always return `available = true` from the lua script directly instead of hard-wiring the go code to return true if the availability of an action is ambiguous.

File: util/lua/lua.go [hunk 1]
```diff
@@ -239,7 +239,7 @@ func (vm VM) ExecuteResourceActionDiscovery(obj *unstructured.Unstructured, scri
 func isActionAvailable(actionsMap interface{}) bool {
 	actions, ok := actionsMap.(map[string]interface{})
 	if !ok {
-		return false
+		return true
```
Comment (simster7) [inline util/lua/lua.go]: Currently, this change would mean that when a `discovery.lua` script returns a malformed or unexpected result, then the action would be displayed as available.

Comment (alexmt) [reply util/lua/lua.go]: We don't have many actions yet, only deployment restart and rollout resume. But in both cases available/not available is UI only feature. The actions scrips are idempotent and work correctly in both available/not available states.

**Comment by jessesuen** [comment]
I think the preferred user experience for writing actions, is for them to be available by default as it requires less knowledge on the resource action author. For example, exposing an action should be as simple as this.

```lua
actions = {}
actions["restart"] = {}
return actions
```

> Yes, @jessesuen and I knew this would have to be done when merging.

I think when I said that, I understood that we would have to change all action scripts such that it always returns all actions in general, while providing a way to denote which ones were enabled/disabled.

Thinking about this some more, I think the best way solve this is to rename the field to: `disable`, which means that the action will be disabled from UI.  The term available might be misnamed because "all actions" and "available actions" could mean the same thing to many people.

**Comment by simster7** [comment]
Sure this makes sense to me.

>  I understood that we would have to change all action scripts such that it always returns all actions in general

You are right: all actions are always returned when running `argocd app actions list` and can be run regardless of whether or not they are "available".

I do now agree that it would be better to make all actions seem available unless a developer chooses to disable them under some circumstances. In this case we are tasking the developer in making sure their actions are disabled when they shouldn't be run, which makes more sense than tasking the users (like Jesse says).

Another point is if we allow developers to disable actions via a `discovery.lua` script, do we want to actively prevent users from running them on the CLI and UI? Or do we task the developers to make that check when running the `action.lua` script?

@alexmt @jessesuen I can own this change if and how we decide to go ahead with it.

**Comment by alexmt** [comment]
@simster7 sure, please feel free to work on this change.

I would propose to rename `available` to `disabled` and actively prevent action execution if discovery lua returns `disabled=true`

**Pull request closed without merging**